primitives/header-chain/src/justification/verification/equivocation.rs

```diff
@@ -177,3 +172,39 @@ impl<'a, Header: HeaderT> JustificationVerifier<Header> for EquivocationsCollect
 		Ok(())
 	}
 }
+
+/// Helper struct for finding equivocations in GRANDPA proofs.
+pub struct GrandpaEquivocationsFinder<C>(sp_std::marker::PhantomData<C>);
+
+impl<C: ChainWithGrandpa>
+	FindEquivocations<
+		GrandpaJustification<HeaderOf<C>>,
+		JustificationVerificationContext,
+		EquivocationProof<HashOf<C>, BlockNumberOf<C>>,
+	> for GrandpaEquivocationsFinder<C>
+{
+	type Error = Error;
+
+	fn find_equivocations(
+		verification_context: &JustificationVerificationContext,
+		synced_proof: &GrandpaJustification<HeaderOf<C>>,
+		source_proofs: &[GrandpaJustification<HeaderOf<C>>],
+	) -> Result<Vec<EquivocationProof<HashOf<C>, BlockNumberOf<C>>>, Self::Error> {
+		let source_proof_idx = match source_proofs
```

Comment:
Since we consider theoretical `1/3 + 1` malicious validators attack (and network split), I'd avoid relying on the fact that the `source_proofs` is ordered by round number here. For the safety, let's do something like:
```rust
for source_proof in source_proofs.iter().filter(|p| p.round == synced_proof.round) {
...
}
```